### PR TITLE
Fix --persist `mila code` bug and intermittent connection errors [MT-78]

### LINKED ISSUE #100: [v0.1.2] `mila code` with `--persist` fails!
### What command did you run?
`mila code repos/office_hours --persist`

### Describe the bug
```console
$ mila code repos/office_hours --persist
(mila) $ lfs quota -u $USER $HOME
Disk quotas for usr normandf (uid 1471600598):
     Filesystem  kbytes   quota   limit   grace   files   quota   limit   grace
/home/mila/n/normandf
                95748040       0 104857600       -  908726       0 1048576       -
uid 1471600598 is using default block quota setting
uid 1471600598 is using default file quota setting
[02/13/24 15:34:18] WARNING  2024-02-13 15:34:18,136 - WARNING - Unable to check the disk-quota on the cluster: not enough   commands.py:534
                             values to unpack (expected 9, got 1)                                                                           
sbatch: error: Unable to open file ~/.milatools/batch/batch-1707856458144355897.sh
touch: cannot touch '.milatools/batch/out-1707856458144355897.txt': No such file or directory
tail: cannot open '.milatools/batch/out-1707856458144355897.txt' for reading: No such file or directory
tail: no files remaining
Traceback (most recent call last):
  File "/home/fabrice/miniconda3/lib/python3.11/site-packages/milatools/cli/commands.py", line 80, in main
    mila()
  File "/home/fabrice/miniconda3/lib/python3.11/site-packages/milatools/cli/commands.py", line 383, in mila
    return function(**args_dict)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/fabrice/miniconda3/lib/python3.11/site-packages/milatools/cli/commands.py", line 572, in code
    data, proc = cnode.ensure_allocation()
                 ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/fabrice/miniconda3/lib/python3.11/site-packages/milatools/cli/remote.py", line 495, in ensure_allocation
    login_node_runner, results = self.extract(
                                 ^^^^^^^^^^^^^
  File "/home/fabrice/miniconda3/lib/python3.11/site-packages/milatools/cli/remote.py", line 364, in extract
    promise.join()
  File "/home/fabrice/miniconda3/lib/python3.11/site-packages/invoke/runners.py", line 1622, in join
    return self.runner._finish()
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/fabrice/miniconda3/lib/python3.11/site-packages/invoke/runners.py", line 518, in _finish
    raise UnexpectedExit(result)
invoke.exceptions.UnexpectedExit: Encountered a bad command exit code!

Command: "cd $SCRATCH && sbatch -J mila-code '~/.milatools/batch/batch-1707856458144355897.sh'; touch .milatools/batch/out-1707856458144355897.txt; tail -n +1 -f .milatools/batch/out-1707856458144355897.txt"

Exit code: 1

Stdout: already printed

Stderr: n/a (PTYs have no stderr)



An error occurred during the execution of the command `code`. Please try updating milatools by running
  pip install milatools --upgrade
in the terminal. If the issue persists, consider filling a bug report at
  https://github.com/mila-iqia/milatools/issues/new?labels=code%2C0.0.18&template=bug_report.md&title=%5Bv0.0.18%5D+Issue+running+the+command+%60mila+code%60
Please provide the error traceback with the report (the red text above).
```

### Desktop (please complete the following information):
 - OS: Ubuntu 22.04

### LINKED ISSUE #102: [v0.1.2] Intermittent connection errors ("ssh-copy-id appears to have failed", "An error occured while trying to establish a connection with mila")
Intermittent errors during `mila init` and other commands:

```
'ssh-copy-id mila' appears to have failed!
```

 ```
 ERROR: An error happened while trying to establish a connection with mila
        -The cluster might be under maintenance
           Check #mila-cluster for updates on the state of the cluster
        -Check the status of your connection to the cluster by ssh'ing onto it.
        -Retry connecting with mila
        -Try to exclude the node with -x mila parameter
 ```
 
 For example:
 
 ```console
 $ mila -vvv init
(...)
Checking connection to compute nodes
 (...)
[02/14/24 14:19:21] ERROR    2024-02-14 14:19:21,979 - ERROR - Exception (client): Error reading SSH protocol banner                                      transport.py:1893
                    ERROR    2024-02-14 14:19:21,985 - ERROR - Traceback (most recent call last):                                                         transport.py:1891
                    ERROR    2024-02-14 14:19:21,988 - ERROR -   File                                                                                     transport.py:1891
                             "/home/fabrice/miniconda3/envs/milatools/lib/python3.11/site-packages/paramiko/transport.py", line 2292, in _check_banner                     
                    ERROR    2024-02-14 14:19:21,991 - ERROR -     buf = self.packetizer.readline(timeout)                                                transport.py:1891
                    ERROR    2024-02-14 14:19:21,993 - ERROR -           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                transport.py:1891
                    ERROR    2024-02-14 14:19:21,994 - ERROR -   File                                                                                     transport.py:1891
                             "/home/fabrice/miniconda3/envs/milatools/lib/python3.11/site-packages/paramiko/packet.py", line 374, in readline                              
                    ERROR    2024-02-14 14:19:21,996 - ERROR -     buf += self._read_timeout(timeout)                                                     transport.py:1891
                    ERROR    2024-02-14 14:19:21,998 - ERROR -            ^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                     transport.py:1891
                    ERROR    2024-02-14 14:19:21,999 - ERROR -   File                                                                                     transport.py:1891
                             "/home/fabrice/miniconda3/envs/milatools/lib/python3.11/site-packages/paramiko/packet.py", line 611, in _read_timeout                         
[02/14/24 14:19:22] ERROR    2024-02-14 14:19:22,001 - ERROR -     raise socket.timeout()                                                                 transport.py:1891
                    ERROR    2024-02-14 14:19:22,002 - ERROR - TimeoutError                                                                               transport.py:1891
                    ERROR    2024-02-14 14:19:22,004 - ERROR -                                                                                            transport.py:1891
                    ERROR    2024-02-14 14:19:22,005 - ERROR - During handling of the above exception, another exception occurred:                        transport.py:1891
                    ERROR    2024-02-14 14:19:22,007 - ERROR -                                                                                            transport.py:1891
                    ERROR    2024-02-14 14:19:22,008 - ERROR - Traceback (most recent call last):                                                         transport.py:1891
                    ERROR    2024-02-14 14:19:22,009 - ERROR -   File                                                                                     transport.py:1891
                             "/home/fabrice/miniconda3/envs/milatools/lib/python3.11/site-packages/paramiko/transport.py", line 2113, in run                               
                    ERROR    2024-02-14 14:19:22,010 - ERROR -     self._check_banner()                                                                   transport.py:1891
                    ERROR    2024-02-14 14:19:22,011 - ERROR -   File                                                                                     transport.py:1891
                             "/home/fabrice/miniconda3/envs/milatools/lib/python3.11/site-packages/paramiko/transport.py", line 2296, in _check_banner                     
                    ERROR    2024-02-14 14:19:22,012 - ERROR -     raise SSHException(                                                                    transport.py:1891
                    ERROR    2024-02-14 14:19:22,013 - ERROR - paramiko.ssh_exception.SSHException: Error reading SSH protocol banner                     transport.py:1891
                    ERROR    2024-02-14 14:19:22,014 - ERROR -                                                                                            transport.py:1891
ERROR: An error happened while trying to establish a connection with mila
        -The cluster might be under maintenance
           Check #mila-cluster for updates on the state of the cluster
        -Check the status of your connection to the cluster by ssh'ing onto it.
        -Retry connecting with mila
        -Try to exclude the node with -x mila parameter
``` 
 
 
 These all seem to be caused by this "Error reading SSH protocol banner" error.

### PR DESCRIPTION
**Pull request by lebrice**
- Fixes #100
- Fixes an unnecessary `cd $SCRATCH` before `salloc` and `sbatch` on the mila cluster
- Fixes  #102
    - Adds a `banner_timeout: 60` in the `connect_kwargs` that are passed from fabric to `paramiko.SSHClient.connect` for the mila cluster when we create a `milatools.cli.remote.Remote` object. This is the most "experimental" change that might require some more testing. In my experience it consistently solved a new paramiko "Unable to parse SSH banner" error that I started experiencing during tests.

### CODE REVIEW

File: milatools/cli/utils.py [hunk 1]
```diff
@@ -149,8 +160,12 @@ def __str__(self):
...
             + "-Retry connecting with mila"
             + "\n\t"
-            + f"-Try to exclude the node with -x {self.node_hostname} "
-            "parameter"
+            + f"-Try to exclude the node with -x {self.node_hostname} parameter\n"
+            + "\n"
+            + "You might want to also include this detailed error message:\n"
+            + "\n```\n"
+            + str(self.error)
+            + "\n```\n"
```
Comment (satyaog) [inline milatools/cli/utils.py:168]:
```suggestion
            + "If you reach out for help, you might want to also include this detailed error message:\n"
            + "\n```\n"
            + str(self.error)
            + "\n```\n"
```

File: milatools/cli/commands.py [hunk 2]
```diff
@@ -618,6 +617,8 @@ def code(
                 "The editor was closed. Reopen it with <Enter>"
                 " or terminate the process with <Ctrl+C>"
             )
+            if "pytest" in sys.modules:
+                break
```
[the remark this answers is not in the document]
Comment (lebrice) [reply milatools/cli/commands.py:621]: This just checks if it's run from within pytest. This doesn't change anything during normal operation.

File: milatools/cli/commands.py [hunk 3]
```diff
@@ -1074,54 +1075,83 @@ def _standard_server(
...
+        ):
+            header_line = line
+            header_line_index = index
+            break
+    assert header_line
+    assert header_line_index
+
+    values_line_parts: list[str] = []
+    # The next line may overflow to two (or maybe even more?) lines if the name of the
+    # $HOME dir is too long.
```
[the remark this answers is not in the document]
Comment (lebrice) [reply milatools/cli/commands.py:1099]: Words aren't split, but what happens is that the home folder is on a single line and the rest on another line.
This here would also work the quotas somehow needed more than one line to display, also.

File: milatools/cli/commands.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lebrice) [reply milatools/cli/commands.py:621]: It's not leftover from debug, it's actually just an easy way to change the behaviour of the code when it's being run within pytest. I'll make a small function to make this easier to read.

[the remark this answers is not in the document]
Comment (lebrice) [reply milatools/cli/commands.py]: Addressed in 6411d300e7725793087dafa6346269e8697ffe1e

**Pull request merged**